File: matlab/src/Sim.m
function [params,data] = Sim(params)

% Initialize grids, distribution functions and output array (data)
[params, fs, data] = initialize_simulation(params);

% Plot initial conditions
params.Efield = vPoisson(fs,params.grids,params.charge);
params.Efield_list(:,1) = params.Efield;

% first plot the initial condition
plot_results(params, fs);

% Main loop over time
Nsamples = 0;
for iT = 1:params.Nt_max
    params.it = iT;
    % Perform a single time step
    [fs, params] = step(params, fs);
    
    % increase time
    time = params.dt * iT

    % Measurements
    [params]=measure(params, fs);

    % Plot results at each time step
    plot_results(params, fs);

    % save config at specific times
    if mod(iT,params.dit_save) == 0
        Nsamples = Nsamples + 1;
        data = save_config(params,data,fs,Nsamples);
    end

    % Check if simulation end time is reached
    if time >= params.Tend
        break;
    end
end

% save config to file
save_config(params,data,fs,Nsamples,1);

end
%% Helper Functions

function [fs, params] = step(params, fs)


    if params.method == "predcorr"
        [fs,params] = predictor_corrector(params,fs);
    elseif params.method == "predcorr_multi"
        [fs,params] = predictor_corrector_subcycling_electrons(params,fs);
    elseif params.method == "NuFi"
        [fs,params] = NuFi(params,fs);
    elseif params.method == "CMM"
        [fs,params] = CMM(params,fs);
    else
        display("error step")
    end


end

function [fs,params] = CMM(params,fs)
iT = params.it+1;
dt = params.dt;
for s = 1:params.Ns
    [X,V] = sympl_flow_Half(iT,dt,params.grids(s).X,params.grids(s).V,params.charge(s)/params.Mass(s)*params.Efield_list,params.grids(s));
    fini = params.fini{s};
    fs(:,:,s) = fini(X,V);
end
[Efield] =vPoisson(fs,params.grids,params.charge);
params.Efield = Efield;
params.Efield_list(:,iT) = Efield;


end


function plot_results(params, fs)
    % Plot results at each time step
    for s = 1:params.Ns
    subplot(params.Ns+1, 1, s);
    pcolor(params.grids(s).X, params.grids(s).V, fs(:, :, s)); shading flat;
    subtitle("$f_\mathrm{"+params.species_name(s)+"}$");
    colorbar();
    xlabel("$x$");
    ylabel("$v$");
    end
   
    subplot(params.Ns+1, 1, params.Ns+1);
    plot(params.grids(1).x, params.Efield); xlim([params.grids(1).x(1), params.grids(1).x(end)]);
    subtitle("$E$");
    colorbar();
    xlabel("$x$");

    pause(0.01); % Pause for visualization
end

File: python/src/Sim.py
import numpy as np
import matplotlib.pyplot as plt

import initialize_simulation

def Sim(params):
    """
    Runs the plasma simulation.
    
    Args:
        params (dict): Dictionary containing simulation parameters.
    
    Returns:
        params (dict): Updated parameters.
        data (dict): Simulation results.
    """

    # Initialize grids, distribution functions and output data
    params, fs, data = initialize_simulation(params)

    # Compute initial electric field and store it
    params.Efield = vPoisson(fs, params["grids"], params["charge"])
    params["Efield_list"] = [params["Efield"]]  # Store first Efield

    # Plot initial conditions
    plot_results(params, fs)

    # Main time loop
    Nsamples = 0
    for iT in range(1, params["Nt_max"] + 1):
        params["it"] = iT  # Store iteration count

        # Perform a single time step
        fs, params = step(params, fs)

        # Increase time
        time = params["dt"] * iT

        # Measurements
        params = measure(params, fs)

        # Plot results
        plot_results(params, fs)

        # Save configuration at specific times
        if iT % params["dit_save"] == 0:
            Nsamples += 1
            data = save_config(params, data, fs, Nsamples)

        # Stop if simulation end time is reached
        if time >= params["Tend"]:
            break

    return params, data


def step(params, fs):
    """
    Advances the simulation by one time step.
    """
    method = params["method"]
    
    if method == "predcorr":
        fs, params = predictor_corrector(params, fs)
    elif method == "NuFi":
        fs, params = NuFi(params, fs)
    elif method == "CMM":
        fs, params = CMM(params, fs)
    else:
        print("Error: Unknown step method")

    return fs, params


def CMM(params, fs):
    """
    Runs the CMM (Conservative Mapping Method) step.
    """
    iT = params["it"] + 1
    dt = params["dt"]

    for s in range(params["Ns"]):
        X, V = sympl_flow_Half(
            iT, dt, params["grids"][s]["X"], params["grids"][s]["V"],
            params["charge"][s] / params["Mass"][s] * params["Efield_list"], 
            params["grids"][s]
        )
        fini = params["fini"][s]
        fs[:, :, s] = fini(X, V)

    params["Efield"] = vPoisson(fs, params["grids"], params["charge"])
    params["Efield_list"].append(params["Efield"])

    return fs, params


def plot_results(params, fs):
    """
    Plots the results of the simulation.
    """
    fig, axes = plt.subplots(params["Ns"] + 1, 1, figsize=(8, 6))

    for s in range(params["Ns"]):
        ax = axes[s]
        im = ax.pcolormesh(params["grids"][s]["X"], params["grids"][s]["V"], fs[:, :, s], shading='auto')
        ax.set_title(f"$f_\\mathrm{{{params['species_name'][s]}}}$")
        fig.colorbar(im, ax=ax)
        ax.set_xlabel("$x$")
        ax.set_ylabel("$v$")

    ax = axes[-1]
    ax.plot(params["grids"][0]["x"], params["Efield"])
    ax.set_title("$E$")
    ax.set_xlabel("$x$")
    fig.colorbar(im, ax=ax)

    plt.pause(0.01)  # Pause for visualization
    plt.show(block=False)  # Non-blocking display

